pycaw/magic.py

```diff
@@ -0,0 +1,563 @@
+                    self.magic_app.volume_callback(new_volume)
+
+            if (self.magic_session and
+                    changer_guid != self.magic_session.guid.contents):
+
+                if self.magic_session.volume_callback:
+                    self.magic_session.volume_callback(new_volume)
+
+        # check old mute vs new:
+        elif self.mute != new_mute:
+            # self.mute will keep the none state
+            # until self._activate()
+            self.mute = new_mute
+            changer_guid = event_context.contents
+
+            # send callbacks, if defined callback exists
+            # and if the mute is new:
+            if (self.magic_app and
+                    changer_guid != self.magic_app.guid.contents and
+                        self.magic_app.mute_callback):
+
+                self.magic_app.mute_callback(new_mute)
+
+            if (self.magic_session and
+                    changer_guid != self.magic_session.guid.contents
+                        self.magic_session.mute_callback):
+
+                self.magic_session.mute_callback(new_mute)
+
+    def OnStateChanged(self, new_state_id):
+        new_state = self.AudioSessionState[new_state_id]
+
+        self.state = (new_state, new_state_id)
+        self.state_id = new_state_id
+
+        if new_state_id == 3 and self.magic_manager:
+            # if magic_root_session is managed by the magic_manager,
+            # wich is the default for use with MagicApp(("app_exe",))
+            self.magic_manager.remove_session(self.iid,
+                                              self.magic_app)
+
+        # send callbacks, if defined
+        if self.magic_app and self.magic_app.state_callback:
+            self.magic_app.state_callback(new_state, new_state_id)
+
+        if self.magic_session and self.magic_session.state_callback:
+            self.magic_session.state_callback(new_state, new_state_id)
+
+    def _get_app_exe(self):
+        self.pid = self._ctl2.GetProcessId()
+
+        if self.pid != 0:
+            # try:
+            app_exe = psutil.Process(self.pid).name()
+            # except psutil.NoSuchProcess:
+            # for some reason GetProcessId returned an non existing pid
+            # TODO:
+            # that should not happen right?
+        else:
+            # System Sound:
+            # self._ctl2.GetDisplayName() returns:
+            # @%SystemRoot%\System32\AudioSrv.Dll,-202
+            # for system sounds
+            # TODO: Double check? or is only self.pid == 0
+            # for System Sound
+            app_exe = "SndVol.exe"
+        return app_exe
+
+    def register_notification(self):
+        self._ctl2.RegisterAudioSessionNotification(self)
+
+    def unregister_notification(self):
+        self._ctl2.UnregisterAudioSessionNotification(self)
+
+
+# TODO
+# i dont really now how to implement this.
+# this all should be actived once on MagicApp or MagicManager use
+# or already at import of this module?
```

Comment:
Yeah I think ideally we should avoid magic to happen at import time for multiple reasons.
One is it makes it harder to (unit) test for instance

Reply:
yeah i intregated it with .activate_magic()